### ISSUE TITLE
Cannot trainer.test() with checkpoints saved after training a model on multiple GPUs and nodes (using DDP)

### ISSUE THREAD

**Issue by amorehead** [PyTorchLightning/pytorch-lightning#9153]
## 🐛 Bug

Cannot train a model using multiple GPUs and then test it using only a single GPU. That is, I have to use the same environment (e.g., number of GPUs/nodes) in which I trained the model to perform testing on a held-out dataset. I am using DDP as my distributed backend.

### To Reproduce

Train a model using 16 GPUs across 4 nodes on a computer cluster like Summit and then test one of the checkpoints either on the same cluster (in a second call to the training script with max_epochs=$the_last_epoch_completed) or on a local machine in a separate script calling trainer.test().

### Expected behavior

Trainer should be able to initialize the process group such that only 1 GPU is used at inference time, while multiple GPUs can be used to train the model.

### Environment

```
* CUDA:
	- GPU:
		- Tesla V100-SXM2-16GB
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.2
	- pyTorch_debug:     False
	- pyTorch_version:   1.7.1
	- pytorch-lightning: 1.4.4
	- tqdm:              4.62.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         ppc64le
	- python:            3.8.8
	- version:           #1 SMP Thu Feb 18 09:47:51 EST 2021
```

### Additional context

Exact error trace I'm seeing:
```
Using backend: pytorch
Global seed set to 42
DB5 cache found
Loaded DB5 train-set, source: datasets/DB5/final/processed, length: 140
DB5 cache found
Loaded DB5 val-set, source: datasets/DB5/final/processed, length: 35
DB5 cache found
Loaded DB5 test-set, source: datasets/DB5/final/processed, length: 55
/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: Metric `AUROC` will save all targets and predictions in buffer. For large datasets this may lead to large memory footprint.
  warnings.warn(*args, **kwargs)
/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: Metric `AveragePrecision` will save all targets and predictions in buffer. For large datasets this may lead to large memory footprint.
  warnings.warn(*args, **kwargs)
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:446: UserWarning: Checkpoint directory checkpoints exists and is not empty.
  rank_zero_warn(f"Checkpoint directory {dirpath} exists and is not empty.")
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
Traceback (most recent call last):
  File "/home/alexm/Repositories/Lab_Repositories/DeepInteract/project/lit_model_test.py", line 186, in <module>
    main(args)
  File "/home/alexm/Repositories/Lab_Repositories/DeepInteract/project/lit_model_test.py", line 128, in main
    model = model.load_from_checkpoint(ckpt_path)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/pytorch_lightning/core/saving.py", line 131, in load_from_checkpoint
    checkpoint = pl_load(checkpoint_path, map_location=lambda storage, loc: storage)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/pytorch_lightning/utilities/cloud_io.py", line 33, in load
    return torch.load(f, map_location=map_location)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/torch/serialization.py", line 607, in load
    return _load(opened_zipfile, map_location, pickle_module, **pickle_load_args)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/torch/serialization.py", line 882, in _load
    result = unpickler.load()
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 576, in __setstate__
    self.process_group = _get_default_group()
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/torch/distributed/distributed_c10d.py", line 358, in _get_default_group
    raise RuntimeError("Default process group has not been initialized, "
RuntimeError: Default process group has not been initialized, please make sure to call init_process_group.
```

**Comment by amorehead** [comment]
@awaelchli and @tchaton, when I view my checkpoint files in a text editor (Gnome's, for instance), this is what I'm seeing (for its head): 

![Screenshot from 2021-08-28 17-18-00](https://user-images.githubusercontent.com/7051982/131232369-7d85cef2-568c-4146-bd5f-d8b38b0a4072.png)

and its tail:

![Screenshot from 2021-08-28 17-19-44](https://user-images.githubusercontent.com/7051982/131232388-a3570bfe-aa68-4bf1-ac09-33b0e5b23287.png)

**Comment by awaelchli** [comment]
@amorehead is it maybe possible that this only happens with the wandb logger? 

could you maybe share the checkpoint file you are trying to reload? I'll try to have a look inside. You can also try to look inside by initializing the progress group as the error message suggests.

**Comment by amorehead** [comment]
@awaelchli, attached below is one of the checkpoints that has been giving me this error:

[checkpoint.tar.gz](https://github.com/PyTorchLightning/pytorch-lightning/files/7071622/checkpoint.tar.gz)

**Comment by amorehead** [comment]
@awaelchli,

It appears as though the WandBLogger is not the root cause of the issue, at least in my view. I tried swapping it out for the TensorBoardLogger, and I receive the same process group error when I go to call trainer.test().

I was able to load in the checkpoint when initializing the process group manually, as you suggested. I cannot identify any areas in the dictionary where process groups are mentioned, at least on my first few passes, so I am a bit confused as to what is triggering the process group error in the first place.

In addition, I tried running trainer.test() with my (now) loaded-in checkpoint to run inference, and I found that, stemming from my LightningModule's self.save_hyperparameters() call in its init() method, I am seeing a new error. It seems as though my LightningModule is trying to pickle a _SingleProcessDataLoaderIter instance when it saves its hyperparameters.

```
/home/alexm/anaconda3/envs/DeepInteract/bin/python /home/alexm/Repositories/Lab_Repositories/DeepInteract/project/lit_model_test.py --logger_name WandB --experiment_name LitGINI-b4-gl1-n128-e128-il2-i128-Alex-Local-Model-Testing --online --max_hours 1 --max_minutes 55 --ckpt_dir checkpoints --num_gpus 1 --num_compute_nodes 1 --gpu_precision 32 --num_workers 2 --dips_percent_to_use 1.00 --self_loops --pn_ratio 0.1 --use_dgl --num_epochs 50 --patience 5 --batch_size 1 --accum_grad_batches 1 --lr 3e-4 --weight_decay 1e-4 --dropout_rate 0.1 --model_name gini --gnn_layer_type geotran --num_gnn_layers 1 --num_gnn_hidden_channels 128 --num_gnn_attention_heads 8 --num_interact_layers 128 --interact_module_type resnet --num_interact_hidden_channels 128 --num_interact_attention_heads 8 --patch_size 7 --final_layer_bias_value -7.0 --ckpt_name LitGINI-epoch=00-val_bce=0.31.ckpt
Using backend: pytorch
Global seed set to 42
DB5 cache found
Loaded DB5 train-set, source: datasets/DB5/final/processed, length: 140
DB5 cache found
Loaded DB5 val-set, source: datasets/DB5/final/processed, length: 35
DB5 cache found
Loaded DB5 val-set, source: datasets/DB5/final/processed, length: 16
DB5 cache found
Loaded DB5 test-set, source: datasets/DB5/final/processed, length: 55
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: Metric `AUROC` will save all targets and predictions in buffer. For large datasets this may lead to large memory footprint.
  warnings.warn(*args, **kwargs)
/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: Metric `AveragePrecision` will save all targets and predictions in buffer. For large datasets this may lead to large memory footprint.
  warnings.warn(*args, **kwargs)
Traceback (most recent call last):
  File "/home/alexm/Repositories/Lab_Repositories/DeepInteract/project/lit_model_test.py", line 129, in <module>
    main(args)
  File "/home/alexm/Repositories/Lab_Repositories/DeepInteract/project/lit_model_test.py", line 74, in main
    model = LitGINI.load_from_checkpoint(ckpt_path)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/pytorch_lightning/core/saving.py", line 153, in load_from_checkpoint
    model = cls._load_model_state(checkpoint, strict=strict, **kwargs)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/pytorch_lightning/core/saving.py", line 195, in _load_model_state
    model = cls(**_cls_kwargs)
  File "/home/alexm/Repositories/Lab_Repositories/DeepInteract/project/utils/deepinteract_modules.py", line 1883, in __init__
    self.save_hyperparameters()
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/pytorch_lightning/core/mixins/hparams_mixin.py", line 105, in save_hyperparameters
    save_hyperparameters(self, *args, ignore=ignore, frame=frame)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/pytorch_lightning/utilities/parsing.py", line 252, in save_hyperparameters
    obj._hparams_initial = copy.deepcopy(obj._hparams)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 296, in _reconstruct
    value = deepcopy(value, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 205, in _deepcopy_list
    append(deepcopy(a, memo))
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 296, in _reconstruct
    value = deepcopy(value, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 296, in _reconstruct
    value = deepcopy(value, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/copy.py", line 161, in deepcopy
    rv = reductor(4)
  File "/home/alexm/anaconda3/envs/DeepInteract/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 547, in __getstate__
    raise NotImplementedError("{} cannot be pickled", self.__class__.__name__)
NotImplementedError: ('{} cannot be pickled', '_SingleProcessDataLoaderIter')
```

**Comment by awaelchli** [comment]
yes you need to exclude any unpickleable objects that are passed in via the constructor. 
`self.save_hyperparameters(ignore="some_arg_name")`
maybe you are passing in dataloaders

**Comment by awaelchli** [comment]
After trying to load your checkpoint I am sure you have a hyperparameter object pickled, something specific to your project.
It must be an object you have passed in via `__init__`.

**Comment by amorehead** [comment]
@awaelchli, you are absolutely right. I discovered that the issue was caused by my passing command-line arguments from ArgParse directly into my LightningModule's init() method. That is, after converting my ArgParse arguments into a Python dictionary, I was feeding this dictionary in its entirety into the init() method using the unravel syntax (i.e., **dict_args). ArgParse must have picked up a DataLoader instance somehow while it was parsing my provided CLI arguments, so then my LightningModule would try to pickle a DataLoader (which isn't feasible currently).

The issue is fixed on my end, and now I can test all my checkpoints. Thank you very much for your help in troubleshooting this issue!
